### DISCUSSION TITLE
Poll: Attributes or no?

### DISCUSSION THREAD

**Discussion by Ovid** [Perl-Oshun/oshun#22] [Polls]
Initially attributes and a `KIM` syntax were proposed for data checks:

```perl
my $count :of(INT) = 0;
```

There are two reasons for this. The primary reason was the concern that the `my TYPE VARLIST` syntax was already used and P5P might reject repurposing it. However, it now looks like this was worrying about nothing.

The other reason was pushing Perl to standardize on `KIM` syntax, where appropriate. In particular, with a 36-year-old language, we're piling on syntax after syntax. The language is growing large and we wind up with bewildering "TATMWTDI" (There Are Too Many Ways To Do It) and every developer seems to fight for their one preferred way, meaning that new developers to Perl find a bewildering range of alternatives, some of which can have subtly different behaviors (for example, `&foo($bar)` versus `foo($bar)`). Since the `KIM` syntax is already used in Corinna, it seemed a logical thing to do.

That being said, several people looking at the proposal were happy with attributes while others hated them.

Repurposing existing `my TYPE VARLIST` syntax has advantages:

* It's more concise (less punctuation)
* It's more familiar (and thus more welcoming to new developers)
* Thus, less controversial

@druud @happy-barney @perigrin @rabbiveesh @rwp0 @SysPete @tobyink @demerphq

**Comment by Ovid** [comment]
Please note that the above is only for _declarative_ usage of checks:

* `my int $foo;`
* `sub foo($arg :of(INT))`

There have been plenty of discussions about using types in expressions, but I feel that's out of scope for the current MVP unless we find that it has an impact on the design of the MVP.

**Reply by happy-barney** [reply]
advantage of attributes is that this can be only in library (with some new support in core even with acceptable performance) - I can prepare small spec for that to discuss if there will be interest

**Reply by tobyink** [reply]
Yeah, but I don't think it's good to make syntax decisions based on what is most convenient to implement in the short term. It's basically a form of technical debt, saddling potentially millions of future users with a worse syntax because that makes things easier on a handful of implementers over the next year or two.

**Reply by happy-barney** [reply]
@tobyink my opinion is that for this poll we should have also overview where we want to continue with it.

easy implementation with leading token where you don't need to put heuristics to guess what it is (just look at indirect syntax - should perl support proper class literal syntax, it will not be problem) is imho better for mvp implementation so you can focus on checks and not on grammars and conflicts.

**Reply by demerphq** [reply]
@tobyink the counterpoint to what you say is that choosing something that is inconvenient to implement but supposedly superior also reduces (considerably sometimes) the change that it will get implemented at all. Essentially it boils down to a bird in the hand is better than two in the bush (from the hunting perspective).  I have seen more than one proposal for new Perl features that were so unaligned with what was feasible that the proposal never went anywhere.  I dont have enough insight into the trade offs /here/ to say anything specific, but my general inclination is that choosing a mechanism we can use /right now/ is better than choosing one we might be able to use if a large amount of round tuits were invested into the subject.

**Comment by rwp0** [comment]
`my TYPE VARLIST`

This syntax also let's you define multiple variables with a single type.

I think we should stick to what was there before, if we had a room for `TYPE` let's just take advantage of it.

Let's improve Perl, but not try to change's it's existing syntax fundamentally (as with Raku).

Attributes and KIM is awesome in all other aspects (apart from variable and function parameters TYPE's).

**Comment by demerphq** [comment]
On Mon, 3 Jul 2023, 12:00 Ovid, ***@***.***> wrote:

> Initially attributes and a KIM syntax were proposed for data checks:
>
I am not familiar with the term KIM syntax. Is there a reference link I can
read?

Yves

**Reply by happy-barney** [reply]
KIM = keyword identifier modifiers

in this case it's postfix declaration of checks, using attributes:
```
my $var :attribute(CHECK);
```

**Reply by Ovid** [reply]
@demerphq I see that @happy-barney has covered the basics of it. [I first wrote about it here](https://ovid.github.io/articles/language-design-consistency.html#who-the-hell-is-kim).

**Comment by happy-barney** [comment]
just stupid question - why not both? (ie: timtowdi)
- attributes for mpv (in fact I'd pick separated symbol spaces as standalone project - you can see first draft at https://gist.github.com/happy-barney/57fdaee7d86a0bbdc8e96834e5c3c295)
  - attribute syntax will be still necessary when there is not clear context switch, eg `map`
- then two independent directions are possible
  -  provide prefix
  - work on checks

and let user decide which syntax will be using (as we have with prefix / postfix `if` and co)

**Reply by zmughal** [reply]
> let user decide which syntax will be using (as we have with prefix / postfix if and co)

This is how I feel about it. It retains that natural language aspect of Perl's design where you can choose what to make prominent when reading.

**Reply by duncand** [reply]
I agree with let user decide but that this should be limited.  I propose that the KIM syntax be the standard general syntax that supports the full Oshun syntax, but that a small common subset of the syntax has an alternative KMI, specifically the subset involving just a single alphanumeric word or something very short.  So if you're just saying eg "Int" it can be in either form, but for something arbitrarily long it is only supported in KIM syntax.

**Comment by demerphq** [comment]
Ovid it isnt clear to me what trade offs we are dealing with here. If we choose KIM, would it be more usable than if we used type declarations?

Right now as far as I understand it the trade off is that using type declaration would mean we can use one declaration for multiple variables, and if we used KIM we would have to repeat the declaration for each variable. Are there any other trade offs? Can we use type syntax in subroutine argument declarations, or is that restricted to KIM?

My gut feeling we should choose the option that facilitates the maximum usefulness with the least need to modify the internals.

**Reply by happy-barney** [reply]
@demerphq please don't use word type :-) ... these checks aren't types (eg, in later milestones checks will be available at runtime)

there is imho no tradeoff, you can do both:
```
my Int ($foo, $bar);
my ($foo, $bar) :check(Int);
```

as well as
```
my (Int $oo, Int $bar);
my ($foo :check(Int), $bar :check(Int));
```

**Reply by Ovid** [reply]
Agreed with @happy-barney : please don't say "type" 😃 These are "checks".

We could go deep into the rabbit hole of pros and cons of KIM versus a more traditional check declaration. I don't think KIM would always require repeat declarations:

```perl
my ( $foo, $bar ) :of(INT) = 0;
```

However, KIM was chosen primarily because we were concerned about P5P's reluctance with breaking backwards-compatability. If P5P is ok with the `my TYPE VARLIST` syntax, I would prefer that. (Actually, I prefer KIM because I would like to see future-Perl have a more standardized approach to adding new features, but this feature is so important that I think getting buy-in is more important).

The only potential issue might be extending checks beyond declarations. If we add them to expressions, (`if ($foo is INT){...}`), there could be a change needed, but I don't think that would necessarily entail changing the declaration syntax.

**Reply by Ovid** [reply]
Also, for subroutines, there's an option question about how to specify checks on return values:

```perl
sub frobnicate( INT $times, ARRAY[STR] $stuff) LIST[ NUM, STR ] {
    ...
}
```

(We use `LIST` only for return values. `ARRAY` would mean it's returning an arrayref).

**Reply by demerphq** [reply]
I only used the term type in the sense that we have been discussing:

```
my TYPE $x, $y, $z;
```

versus

```
my $x :of(WHATEVER), $y :of(WHATEVER), $z :of(WHATEVER);
```

I totally get that checks are not types. I just didnt have a good name for the first form other than type.

Does

```
my ( $foo, $bar ) :of(INT) = 0;
```

actually work? It looks a bit weird with only one value but two vars. 

@ovid said:
> I prefer KIM because I would like to see future-Perl have a more standardized approach to adding new features, but this feature is so important that I think getting buy-in is more important

I think we choose the one you think is best unless there is some pressing reason not to. Prior to you saying this I would have said "i prefer the `my TYPE $x` form as it feels more intuitive and handles multiple values currently with no problem." But if you think KIM is better future proofed I can understand the desire to use it instead.

**Reply by Ovid** [reply]
@demerphq Yes, in the `t/heterogeneous_decl.t` test, there's this check:

```perl
my ( $scalar, @array, %hash ) :of(INT) = 0;
```

As for `KIM`, the reason I think it's good is because Perl is getting _flooded_ with keywords, but in Corinna, we discovered we only had four:`class`, `method`, `field`, and `role`. Everything else we were packing in were effectively adjectives modifying the former. By using `KIM`, we avoiding adding a bunch of keywords. To a certain extent, extending a variable with a check could be considered an "adjective" that's modifying the "thing" it's on. So `field $x :reader` would be `NOUN PRONOUN ADJECTIVE` and it's all nice, clean and consistent. I would love to see a future for Perl where new, improved features are adopted in a more consistent syntax to make it more friendly to developers coming to the language (`chomp`, `open` and friends can burn in hell, thank you).

I _think_ it's ok to make an exception in this case because so many languages already use a prefix syntax for this and that won't seem uncomfortable.

I just want to get past this so I can take the work that's there and write up a spec. I literally don't understand which is better.

**Reply by Ovid** [reply]
Also, since it was agreed that checks would not live in the namespace, I think the prefix form probably is safe, but I don't have the knowledge here.

**Reply by Ovid** [reply]
Oh, and `my ( $scalar, @array, %hash ) :of(INT) = 0;` works by assigning the 0 to `$scalar`. The array and hash remain empty.

**Reply by happy-barney** [reply]
@Ovid ... prefix is safe from users point of view, it will be terrible to write parser for it (as long as we use LR(1) parser), unless you restrict its capabilities (use only part of perl, use only in specially designed perl code)

As I mentioned earler, I can write easy to read and maintain (depends on perception of easy though) postfix (attribute) parser.

I'll be also more than happy to provide nightmare test cases to anyone who will try to write prefix grammar :-)

**Reply by perigrin** [reply]
I mean a nightmare test case might be useful to weed out behaviors we don’t want to support or don’t want to support. 

Perl’s parsing is already notoriously and provably hard … I’m not sold on “it’ll be hard to implement” being a limiting factor until @leonerd or someone from P5P says “yeah we won’t implement that” for syntax on a must have feature.

**Reply by happy-barney** [reply]
@perigrin just look at test case later in this discussion ... how to treat `Int` there? what will you restrict? 

Another examples:
```
use constant MIN => 10;
check MIN :isa (NUM) where { $_ == 10 };

my Array[MIN, Int] @list; # is MIN check or constant? you must decide at the time when you read 1 token - `MIN`
```

When allowing full syntax, it may become pretty unreadable:
```
my NUM where { $_ == 10 } $foo;
```

Can check be stored in variable? eg:
```
my $foo = declared check (FOO);

$foo->check ($value);
my $foo $bar;
```

**Reply by perigrin** [reply]
The MIN here is ambiguous depending on the grammar of Array[] syntax but only if we say it can allow non-check values (`Array[10]`) then the choice is obvious … but should that be allowed?

**Reply by Ovid** [reply]
> ```
> use constant MIN => 10;
> check MIN :isa (NUM) where { $_ == 10 };
> 
> my Array[MIN, Int] @list; # is MIN check or constant? you must decide at the time when you read 1 token - `MIN`
> ```

Checks are not in the same namespace and don't have access to the lexical scope, so with our current work. `MIN` cannot be a constant. If we wish to allow dynamic checks to be parameterized at runtime, that's a different story. We didn't do that with the current spec because attributes don't allow that.
 
> ```
> my NUM where { $_ == 10 } $foo;
> ```

We have a spec for _ranges_:

```
my NUM[1..inf] $foo = 1; # 1 <= $_
my NUM[10<..20] $bar = get_value(); # 10 < $_ <= 20
my NUM[5..<15.2] $baz = get_value(); # 5 <= $_ < 15.2
```

But we don't have it for a single value, but we could make it work:

```
NUM[==10] # == 10
NUM[<10]  #  < 10
NUM[<=10] # <= 10
NUM[>10]  #  > 10
NUM[>=10] # >= 10
```

Not sure how clumsy that is, though.

> Can check be stored in variable? eg:
> 
> ```
> my $foo = declared check (FOO);
> 
> $foo->check ($value);
> my $foo $bar;
> ```

Currently no.

**Reply by perigrin** [reply]
It's not currently allowed in the `my TYPE VARLIST` syntax either … but that's mostly because TYPE in this case can _only_ be a package name or a constant that resolves to a package name, i.e. a bareword and thus cannot currently be parametrized. We would have to modify the parser to allow more syntax than a single bareword there if we want _anything_ like parameterized types declared on the fly.

**Reply by Ovid** [reply]
@demerphq I should have read more carefully.

```
my ( $foo, $bar ) :of(INT) = 0;
```

No, that cannot work because `$bar` is not assigned a value.

**Comment by happy-barney** [comment]
Prefix test case:
```
sub Int ($x) { return 10 * $x } # whatever internal representation will be
sub foo { 
  no strict;
  # my $x = 10.5; # edited ... should not be in this example
  Int $x = 11.5;
}

ok foo == 115;
```

**Reply by perigrin** [reply]
I’m not a parser but this doesn’t seem ambiguous to me because the statement with Int used a second time isn’t a declaration (no `my`) so Perl wouldn’t be expecting a check expression there. I would expect that line to be equivalent to `Int($x = 11.5)` … no ambiguity but a PR that would be rejected because nobody should write code like this :)

**Reply by Ovid** [reply]
At the present time that is not problematic because we only allow checks on declaration.

**Reply by happy-barney** [reply]
@Ovid oh, I made a mistake with putting `my` line there :-( (forgetting reading twice these days ...)

regarding your comment ... that's what I'm talking about, without overview where we want to continue after mvp and how it will/may interact with another possible features, it's hard to tell from oversimplified examples.

@perigrin
and that's unexpected ambiguity ... what it will be if `Int` is not a function? for C/Java/ .... programmers that line makes perfect sense, yet it will not behave as expected.

But as I said, I'll be happy to write tests to break it

**Reply by perigrin** [reply]
That ambiguity has nothing to do with Oshun though.  It _already_ exists in Perl, and you have to understand Perl's scoping and parsing _already_ to cope with things like this:

```
perl -Mexperimental=class -E'class Int {} sub Int { ... } Int->new()'
Unimplemented at -e line 1.
```
The fact it already exists is why I'd reject a PR without more explicit definitions of what is going on. I'd tell the submitter "you need to spell that second Int as `Int::` and leave a comment as to why, if you don't understand why come find me and we'll talk about the glory of Perl's bareword parsing."

This is (partly?) why @Ovid keeps saying "Checks are not in the same namespace and don't have access to the lexical scope[…]"

**Reply by demerphq** [reply]
> we'll talk about the glory of Perl's bareword parsing

I'm sure you mean "gory details" there. It sure ain't glorious.

**Reply by happy-barney** [reply]
@perigrin yes, that's true and I'm aware of it ... I'm just pointing out that Oshun should not create another cases of such ambiguity (therefor, first focus on what it has to provide, how it will be accessible is secondary problem)

For example, first implementation part should be support for custom symbol spaces (not only for Oshun but available for any module) - WIP at https://gist.github.com/happy-barney/57fdaee7d86a0bbdc8e96834e5c3c295

**Comment by Ovid** [comment]
Also, unless someone here objects, I plan to list everyone in the Toronto keynote as team members on the Oshun project. I'll try to use your github names and real names. If you'd like to be listed differently, or not at all, please let me know.

**Reply by perigrin** [reply]
Does that make us … [Oshun's 8](https://en.wikipedia.org/wiki/Ocean%27s_8)? I'll see myself out.

**Reply by druud** [reply]
AFAICS, there will still always be room for
```
perl -Mv5.36 -Mexperimental=declared_refs -MData::Dump=dd -wE'
  sub Int { dd @_ }
  Int my ($i, \@j, \%k) = (0, [42..44], {one => 1, two => 2});
'
(0, [42, 43, 44], { one => 1, two => 2 })
```
or similar, to set up checked variables/containers in a more keyword-like way.


Which makes me wish a my() that obeys prototype-context, such that a call like
```
Int my ($i, @j, %k);
```
can set a prototype-context on the my().
But then Int() needs a prototype of `\[$@%&*] ...` for each of its unknown number of parameters, which AFAIK is not offered by prototype. Then this already requires Int to be set up as a keyword.

**Comment by duncand** [comment]
I prefer KIM syntax for the consistency with key-value pairs, where the name always appears on the left of what it is naming, whether it is named arguments or named hash keys.  Also it is consistent with things like SQL where the field name is declared before the possibly numerous descriptors of it.

Also, importantly, KIM syntax is consistent with Corinna field declarations.  If you look at it in a more abstract sense, the check declaration is functionally like many other things that are currently attributes in Corinna field declarations.

And KIM syntax I prefer visually because for the general case where check definitions may be long, it looks better for the long thing to be on the right, its like an arbitrary value expression.

**Reply by zmughal** [reply]
I also prefer the postfix KIM syntax for this reason because it keeps the syntax open to extensibility, but I don't yet know what this would be. I guess on example from what I do with various `/Moo.*/` wrappers around `Getopt` like `MooX::Options` or `CLI::Osprey` is that I have the `isa` next to the `format` coming from the command line:

```perl
option repo_path => (
	is       => 'ro',
	doc      => 'Path to repository',
	format   => 's',
	isa      => AbsDir,
	coerce   => 1,
	required => 0,
	default  => sub { path('.') },
);
```

I might prefer this over looking back and forth, but this is more a developer ergonomics question.

---

The other thing about the KIM syntax would be that if there is ever a syntax to allow using the initial rvalue to indicate what the check would be like the C++ `auto` (and soon to be in C23 afaik), the check would be closer to the rvalue so visually scanning over the code would be over a smaller distance.

**Comment by zmughal** [comment]
Another thing to think about is how the syntax would look for defining checks with destructuring assignment.

**Reply by Ovid** [reply]
Agreed. To do that, we need to see Ingy's full proposal for destructuring assignments.

**Reply by Ovid** [reply]
As a first pass, I would imagine that it would look similar to signatures:

```perl
my [INT $foo, OBJ[My::Customer] @customers, ARRAY[STR] $filters ] = $customer_records;
```

**Comment by esabol** [comment]
Using KIM syntax seems sensible to me, especially since Corinna uses it also.

**Comment by oldtechaa** [comment]
I'm a big fan of using the my TYPE IDENTIFIER syntax. It looks good and it's clean. It doesn't involve a syntax that'll be unfamiliar to nearly everyone coming from other languages, and as mentioned it allows multiple declarations in one statement. Trying to use params to implement a forward-thinking project seems like it'll hold Oshun back quite a lot.

**Reply by happy-barney** [reply]
FYI: as mentioned elsewhere popular many modern languages uses postfix syntax :-)

and found interesting thread on reddit as well https://www.reddit.com/r/AskProgramming/comments/ql2nzj/why_most_new_programming_languages_use_postfix/

**Reply by oldtechaa** [reply]
That is an interesting thread. I guess languages I'm most involved in are either semi-typeless and/or I'm just used to C-style typing, but that may not be the case for all. Maybe for me it goes back to the fact that Perl probably borrows most heavily from C.

**Reply by zmughal** [reply]
Another post I came across on Mastodon: <https://benhoyt.com/writings/name-before-type/>.

**Reply by happy-barney** [reply]
hm, maybe it will be good to summarize these links in primary discussion comment or in wiki (should we enable it here?)

**Comment by druud** [comment]
I presume that at some future point we will also want an `:auto`, that will follow the initial value, with `:any` as an escape.

And then some way to lexically force `:auto` on all underspecified declarations.

**Reply by happy-barney** [reply]
as well as
```
my Some_Check $foo ...;
my same_checks_as ($foo) $bar ...;
```

**Comment by zmughal** [comment]
A quick and dirty search using `grep.metacpan.org` for use of both kinds of syntax:

- [(my|our|state) TYPE VAR](https://grep.metacpan.org/search?p=1&q=%5E%5Cs%2A%28my%7Cour%7Cstate%29%5Cs%2B%2B%28%3F%21sub%7C%3D%3E%29%5B%5E%5Cs%24%25%40%28%5C%5C%5D%2B%2B%5Cs%2B%2B%28%3F%3D%5B%24%25%40%28%5C%5C%5D%29&qft=%2A.pm)

- [attributes on variables and subs](https://grep.metacpan.org/search?size=20&_bb=122719375&q=%5E%5Cs*%28my%7Cour%7Cstate%29%5Cs%2B%2B%28%5C%28%5Cs%2B%29%3F%5B%24%25%40%5C%5C%5D%5B%5E%3D%3B%5D*%3A%7C%28%3F%3C%21%5B%24%40%25%5D%29%5Cbsub%5Cb%5B%5E%7B%5D*%5Cs%2B%3A%5Cs%2B%5Cw%2B&qd=&qft=*.pm&qifl=)

**Reply by perigrin** [reply]
So the first one is going to be exclusively people who are looking to annotate with constraints, while the second is all of Catalyst and anything that uses attributes which is currently almost never a constraint check … so I would argue that this leans away from KIM _in this case_ because there are so many instances in the second result set.

**Reply by happy-barney** [reply]
I think new features should stick with KIM as much as possible ... we are not creating features for past but for future.

**Reply by perigrin** [reply]
I don’t disagree with that. I disagree that this is a _new_ feature considering it’s had valid syntax since 1995. :)

**Reply by happy-barney** [reply]
valid but unused ... and things changed
Perl is not "in" again, we need to be creative, no too conservative

**Reply by perigrin** [reply]
The results above suggest that KIM syntax here _is_ the conservative choice because it’s used for so many other behaviors as well.

@leonerd said somewhere effectively that a constraint declaration should stand out. I honestly have no cat in this fight, and won’t complain if we end up with attribute based check annotations. However, I don’t want to ignore almost 30 years of history for consistency with KIM. Especially if it lessens the impact of the behavior. Finally making a reasonable implementation of a feature that has been 30 years in the making … that’s a compelling story about Perl being dynamic and forward looking without throwing out the past.

**Reply by happy-barney** [reply]
regarding "things changed " and this project:
- it itself counts with separated symbol spaces (with extended grammar) so it will make conflict in parser/lexer lookup
- `my` is not used only for variables anymore (`my sub`), with possible more (can imagine eg `my package ...`, `my attribute ...`, `my constant ...`)

**Reply by leonerd** [reply]
A reminder once again that the parsing rules inside attribute values *are different* than the parsing rules elsewhere in perl. Any attribute value is treated just like a simple string - albeit one where we count the parens and ensure they're balanced. And that is it. It's a string. If you wanted it to be complicated perl code then someone has to `eval()` that, and that gets complicated. It also complicates the parse model.

To show what I mean, here's a really contrieved example wherein I want to restrict a string variable to not permit an open paren inside its value; done as either a leading type constraint, or an attribute.

```perl
my RestrictedString(qr/[^(]/) $str;

my $str :Constrained(RestrictedString(qr/[^(]/));
```

The first one, while awkward and complicated, at least has a *hope* of parsing correctly. The latter one is straight up outright a parse error - because the parens are unbalanced.

This is the reason why I am *strongly* in favour of not really using attributes for this. I'm using them right now today in my `:Checked` field attribute for Object::Pad because there is no better option right now. But a design process can design itself a better way. Please do so.

**Reply by happy-barney** [reply]
@leonerd aware of that and working that on proposal to handle this as well as part of separated symbol spaces (hope to finish it coming weekend)
